openprescribing/matrixstore/management/commands/matrixstore_init.py

```diff
@@ -0,0 +1,131 @@
+        -- matrices of shape (number of practices, number of months)
+        value BLOB,
+
+        PRIMARY KEY (name)
+    );
+
+    -- Maps each practice code to its corresponding row offset in the data matrix
+    CREATE TABLE practice (
+        offset INTEGER,
+        code TEXT UNIQUE,
+
+        PRIMARY KEY (offset)
+    );
+
+    -- Maps each date to its corresponding column offset in the data matrix
+    CREATE TABLE date (
+        offset INTEGER,
+        date TEXT UNIQUE,
+
+        PRIMARY KEY (offset)
+    );
+"""
+
+
+class Command(BaseCommand):
+    help = __doc__
+
+    def add_arguments(self, parser):
+        parser.add_argument('end_date', help='YYYY-MM format')
+        parser.add_argument('sqlite_path')
+
+    def handle(self, end_date, sqlite_path, **kwargs):
+        matrixstore_init(end_date, sqlite_path)
+
+
+def matrixstore_init(end_date, sqlite_path):
+    if os.path.exists(sqlite_path):
+        raise RuntimeError('File already exists at: ' + sqlite_path)
+    sqlite_path = os.path.abspath(sqlite_path)
+    temp_filename = get_temp_filename(sqlite_path)
+    sqlite_conn = sqlite3.connect(temp_filename)
+    bq_conn = Client('hscic')
+    sqlite_conn.executescript(SCHEMA_SQL)
+    dates = generate_dates(end_date)
+    import_dates(sqlite_conn, dates)
+    import_practices(bq_conn, sqlite_conn, dates)
+    import_presentations(bq_conn, sqlite_conn)
+    sqlite_conn.commit()
+    sqlite_conn.close()
+    os.rename(temp_filename, sqlite_path)
+
+
+def import_dates(sqlite_conn, dates):
+    sqlite_conn.executemany(
+        'INSERT INTO date (offset, date) VALUES (?, ?)',
+        enumerate(dates)
+    )
+
+
+def import_practices(bq_conn, sqlite_conn, dates):
+    # We only create practice entries for practices which have prescribed in
+    # the current date range.  The stored matrices are sparse, so having
+    # "empty" practice rows doesn't make much difference there. But when we
+    # start summing and processing these we use dense matrices and so it's
+    # better to cut down the number of rows to a minimum.
+    sql = (
+        """
+        SELECT DISTINCT practice FROM {hscic}.prescribing
+          WHERE month >= '%s' AND month <= '%s'
+          ORDER BY practice
+        """
+    )
+    result = bq_conn.query(sql % (min(dates), max(dates)))
+    practice_codes = [row[0] for row in result.rows]
+    sqlite_conn.executemany(
+        'INSERT INTO practice (offset, code) VALUES (?, ?)',
+        enumerate(practice_codes)
+    )
+
+
+def import_presentations(bq_conn, sqlite_conn):
+    # Unlike with practices above, it costs very little to have BNF codes in
+    # the database which are not prescribed against. And we don't actually know
+    # which codes are and aren't used until we apply the "BNF map" which
+    # translates old codes into new codes.
+    result = bq_conn.query(
+        'SELECT bnf_code, is_generic, adq_per_quantity, name'
+        '  FROM {hscic}.presentation'
+        '  ORDER BY bnf_code'
```

Comment:
Multiline strings?  Makes copying/pasting easier.

Reply:
Yes, I should do that